api/notd/token_manager.py

```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
+        startDate = date_util.datetime_from_now()
+        latestUpdate = await self.retriever.get_latest_update_by_key_name(key='token_attributes')
+        latestProcessedDate = latestUpdate.date
+        logging.info(f'Finding changed tokens since {latestProcessedDate}')
+        updatedTokenMetadatasQuery = (
+            TokenMetadatasTable.select()
+            .with_only_columns([TokenMetadatasTable.c.registryAddress, TokenMetadatasTable.c.tokenId])
+            .where(TokenMetadatasTable.c.registryAddress == address)
+            .where(TokenMetadatasTable.c.updatedDate >= latestProcessedDate)
+        )
+        updatedTokenMetadatasQueryResult = await self.retriever.database.execute(query=updatedTokenMetadatasQuery)
+        updatedTokenMetadatas = set(updatedTokenMetadatasQueryResult)
+        logging.info(f'Scheduling processing for {len(updatedTokenMetadatas)} tokens in collection {address}')
+        messages = [UpdateCollectionTokenAttributesMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message() for (registryAddress, tokenId) in updatedTokenMetadatas]
+        await self.tokenQueue.send_messages(messages=messages)
+        await self.saver.update_latest_update(latestUpdateId=latestUpdate.latestUpdateId, date=startDate)
+
+    async def update_collection_token_attributes_deferred(self, registryAddress: str, tokenId: str) -> None:
+        await self.tokenQueue.send_messages(messages=UpdateCollectionTokenAttributesMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message())
+
+    async def update_collection_token_attributes(self, registryAddress: str, tokenId: str) -> None:
+        retrievedTokenAttributes: List[RetrievedTokenAttribute] = await self.tokenAttributeProcessor.get_token_attributes(registryAddress=registryAddress, tokenId=tokenId)
+        tokenAttributes = await self.retriever.list_token_attributes(
+            fieldFilters=[
+                StringFieldFilter(fieldName=TokenAttributesTable.c.registryAddress.key, eq=registryAddress),
+                StringFieldFilter(fieldName=TokenAttributesTable.c.tokenId.key, eq=tokenId),
+            ]
+        )
+        async with self.saver.create_transaction() as connection:
+            if tokenAttributes:
+                existingTokenAttributesTuples = {(existingTokenAttribute.registryAddress, existingTokenAttribute.tokenId, existingTokenAttribute.name): existingTokenAttribute.tokenAttributeId for existingTokenAttribute in tokenAttributes }
+                tokenAttributeIdsToDelete = (list(existingTokenAttributesTuples.values()))
+                logging.info(f'Deleting  {len(tokenAttributeIdsToDelete)} attributes for registryAddress: {registryAddress}, tokenId: {tokenId}')
+                await self.saver.delete_token_attributes(connection=connection, tokenAttributeIds=tokenAttributeIdsToDelete)
+                #NOTE Didn't use comprehension cause connection kept closing
```

Comment:
formoat these properly: `# NOTE(username): message` and `# TODO(username): message`